### ISSUE TITLE
How to manage Vault moving forward? 

### ISSUE THREAD

**Issue by dystewart** [OCP-on-NERC/operations#9]
I'm opening this issue as an extension of  [https://github.com/OCP-on-NERC/nerc-ocp-config/issues/35](https://github.com/OCP-on-NERC/nerc-ocp-config/issues/35).

Once [https://github.com/OCP-on-NERC/nerc-ocp-config/pull/77](https://github.com/OCP-on-NERC/nerc-ocp-config/pull/77) is merged we can test that the backup script is working. When we can confirm this we will be able to decide how we'd ultimately like to have vault configured and managed. 

We have 2 different paths:

1. Continue moving forward with manual vault deployment
-   If we choose this path there is no external configuration of vault available 
-  We'd want to find a way to track the internal vault configuration. The short answer to how Operate First does this is they don't because there's no way around manually creating and updating something like a text file or markdown file
- We'd need to handle auto unseal of the vault. This means either enabling vault's auto unseal feature or create a script to unseal the vault on our own 

2. Move our integration to the banzaicloud vault operator 
- The issue preventing me from utilizing this in the first place was that it doesn't work out of the box from Operator Hub. With some tweaks I have finally had success deploying the operator (though I still need to test it against all the tooling we are using with vault currently)
- It would allow for the configuration and management of vault (policies, secret engines, etc) via gitops through a vault crd. This would solve our problem of documenting our vault config
- There are also auto unseal tools that this operator provides 

Based on discussion with @jtriley a while back I think ultimately we'd prefer the operator path, but what do we think? I will continue testing the operator locally to ensure that it plays nice with ESO and I can explore the auto unseal options if we think the operator is preferred

**Comment by larsks** [comment]
> The issue preventing me from utilizing this in the first place was that it doesn't work out of the box from Operator Hub

Are the issues specific to OpenShift (i.e., it would work out-of-the-box on vanilla Kubernetes)? Or is the operator just broken?

**Comment by dystewart** [comment]
@larsks No the issues are not OpenShift specific, the operator itself is broken
Here is more some more info about the brokeness of the OperatorHub deployment from an issue I opened with BanzaiCloud a while ago: [https://github.com/banzaicloud/bank-vaults/issues/1474](https://github.com/banzaicloud/bank-vaults/issues/1474)
TLDR: The OperatorHub uses an older version of the banzaicloud vault operator, which attempts to deploy resources with deprecated apis. (This is a persistent issue)

The folks from banzaicloud suggested installing via helm: [https://github.com/banzaicloud/bank-vaults/tree/master/charts/vault-operator](https://github.com/banzaicloud/bank-vaults/tree/master/charts/vault-operator)
Interestingly, installing via helm also does not install all the necessary resources. These resources need to be deployed manually after installing the operator with helm:
- vault `crd`
- operator rbac
- `serviceaccount` 
- vault `cr`
- least privileged `scc` which is essential for the `serviceacount` to function properly 

Once all the ancillary resources are deployed (along with the helm deployment of the operator) finally the vault `sts` will finally deploy

**Comment by dystewart** [comment]
Another potential win with this operator is potentially greatly simplifying the vault backup process: [https://banzaicloud.com/blog/vault-backup-velero/](https://banzaicloud.com/blog/vault-backup-velero/)

I am experimenting with this setup

**Comment by dystewart** [comment]
After lots of research into the banzaicloud vault operator I think it makes the most sense to stay the course with our current vault setup. There are many pitfalls in setting up the vault cluster with the operator (I still haven't had success in deploying) and not to mention the operator tries to deploy the vault cluster with an older version of vault (which fails to deploy) and trying to use a newer image also fails. It's worth keeping an eye on for the future if they figure out some of the issues but it's not worth the headache now. Not to mention the OperatorHub banzaicloud doesn't even deploy the operator correctly out of the box.

That being said there is another potential option which could codify some of our vault config in theory, the vault-config-operator which I may take a look at in the meantime.

**Comment by dystewart** [comment]
Update vault config operator also appears to be more work than it's worth getting up and running. 

Having said that I'm going to create a script to automatically unseal the vault which will come in handy in the event of an unexpected node reboot or really any other time the vault becomes sealed unexpectedly. @jtriley @naved001 @larsks last I was aware the keys were being stored in the MOC aws account. Is the intention to keep them there moving forward? If not we should change the location asap so I can configure the script accordingly. If not I will need the location/access to where they are currently being stored

**Comment by dystewart** [comment]
Closing this based on discussion in weekly NERC meeting, (this was mostly a discussion thread and the vault operators are not working with our version of openshift)
